// eldr/lib/language-result.js
/*
Copyright 2023 Nito T.M.
License https://www.apache.org/licenses/LICENSE-2.0 Apache-2.0
Author Nito T.M. (https://github.com/nitotm)
Package npmjs.com/package/eld
*/
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _LanguageResult_results, _LanguageResult_langCodes, _LanguageResult_numNgrams;
import ISO6391 from "iso-639-1";
import { avgScore } from "./avg-score.js";
export class LanguageResult {
    constructor(iso639_1, results, numNgrams, langCodes) {
        _LanguageResult_results.set(this, void 0);
        _LanguageResult_langCodes.set(this, void 0);
        _LanguageResult_numNgrams.set(this, void 0);
        this.iso639_1 = iso639_1;
        __classPrivateFieldSet(this, _LanguageResult_results, results, "f");
        __classPrivateFieldSet(this, _LanguageResult_langCodes, langCodes, "f");
        __classPrivateFieldSet(this, _LanguageResult_numNgrams, numNgrams, "f");
    }
    get languageName() {
        return ISO6391.getName(this.iso639_1);
    }
    // backwards compatibility
    get language() {
        return this.iso639_1;
    }
    /**
     * Converts internal multi-array results, with integer language codes, to final object with ISO 639-1 codes
     */
    getScores() {
        const scores = {};
        for (const result of __classPrivateFieldGet(this, _LanguageResult_results, "f")) {
            const score = result[1];
            if (score === 0) {
                break;
            }
            scores[__classPrivateFieldGet(this, _LanguageResult_langCodes, "f")[result[0]]] = score;
        }
        return scores;
    }
    isReliable() {
        if (__classPrivateFieldGet(this, _LanguageResult_results, "f").length === 0 || __classPrivateFieldGet(this, _LanguageResult_numNgrams, "f") < 3) {
            return false;
        }
        const nextScore = __classPrivateFieldGet(this, _LanguageResult_results, "f").length > 1 ? __classPrivateFieldGet(this, _LanguageResult_results, "f")[1][0] : 0;
        // A minimum of a 24% per ngram score from average
        return !(avgScore[this.iso639_1] * 0.24 > __classPrivateFieldGet(this, _LanguageResult_results, "f")[0][1] / __classPrivateFieldGet(this, _LanguageResult_numNgrams, "f") ||
            0.01 > Math.abs(__classPrivateFieldGet(this, _LanguageResult_results, "f")[0][1] - nextScore));
    }
}
_LanguageResult_results = new WeakMap(), _LanguageResult_langCodes = new WeakMap(), _LanguageResult_numNgrams = new WeakMap();
